### ISSUE TITLE
Support multiple user settings

### ISSUE THREAD

**Issue by smsegal** [zed-industries/zed#20038]
### Check for existing issues

- [X] Completed

### Describe the feature

It would be useful to support an `includes` directive in the settings json file. 
Particularly for the use-case of sharing configuration across different machines. 

For example, with the Remote Development feature, I now have a bunch of work servers and project paths in my config file, which I share and sync across my work and personal machines. 

If we could conditionally include other config files to extend the "base" config, I would be able to continue sharing my configuration between work/personal machines without irrelevant settings being shared. 

I would imagine that this could work in a few ways: 

1. An "includes" key that takes a list of (relative?) paths to other files to include
2. The same "includes" key, but could take a list of objects that might specify what to do if the file doesn't exist. 

The settings in the included configurations should likely take precedence over all previous settings in a LIFO manner - such that if I include files `A, B, C` that include the same keys `K`, the values should be merged with `C` taking precedence over `B` and so on. 

### If applicable, add mockups / screenshots to help present your vision of the feature

Example settings file: 
```jsonc
{
  # ... existing configuration
  "includes": [
   "work_settings.json",
  ],

  # or maybe
  "includes": [
	{
	  "path": "./work_settings.json",
      "missing": "ok", # or "error",  etc
    }
  ]
}
```

Where "work_settings.json" might include: 

```jsonc
{
  "ssh_connections": [
    {
      "host": "work-server",
      "projects": [
        {
          "paths": [
            "~/project"
          ]
        },
    }
  ]
}
```

**Comment by notpeter** [comment]
See also:
- https://github.com/zed-industries/zed/issues/16294

**Comment by lentzi90** [comment]
Yes please! I want to add to the use-cases also dynamic environments.

I want to keep my settings in general in a dotfiles repo that I pull to any new environment to automatically get the same configuration. However, `ssh_connections` changes so frequently that I do not consider them "settings" in this sense and having them entangled with the version controlled settings in the dotfiles repo is not nice for this reason.

Why do the connections change so often? Because I use temporary test VMs and containers. I create them on demand when I need them and then destroy them as soon as I am done to keep each environment as clean as possible.
I even go as far as using development containers locally through https://devpod.sh/. That is a lot of ssh connections and it would be very nice to keep them separate from the rest of the settings.

**Comment by notpeter** [comment]
I'm not a fan of anything like #include directives, but there are few ways these use cases would be possible to support:

1. `~/.config/zed/settings.d/*.json` (e.g. applied as overrides, alphabetical order of filename?)
2. Specifying an explicit settings.json at runtime -- e.g. `zed --config x/z/settings.json` (this would not support running two settings at the same time)

A real issue for the practicality of any potential solution is the fact that settings.json currently includes a mix of "settings" (available_models, auto_update, ui_font_size, etc) which rarely change and various bits of state which are persisted via settings.json but can also be directly manipulated via Zed actions (theme, ssh_servers, default_model, vim_mode, base_keymap).

As soon as you have multiple files it's unclear where the latter stuff should go if you change it in the UI. In the `work_settings.json` example above, it would hard for Zed to know that editing certain servers should be in one file, but others would be in another (remember, you merged the configs based on order).  Which also leads to the other issue, which is the complexity of merging multiple json files.  Simple keys are easy (decide some order, and deterministically apply multiple files) but things get much harder with merging arrays.  Sometimes you want to merge the arrays, sometimes order matters, sometimes it doesn't, do you need to dedupe values? sometimes!; and sometimes you shouldn't merge. Also applies at arbitrary JSON depths (lsp settings, etc). 

Open to design ideas here, but nothing obvious jumps out at me to address all the desired use cases.

See also: 
- https://github.com/zed-industries/zed/issues/22612

**Comment by yordis** [comment]
I would like to also have a way to share `ssh_connections`, and settings that change often like LLM models, or Context Servers, etc.

**Comment by greg-hellings** [comment]
> I'm not a fan of anything like #include directives, but there are few ways these use cases would be possible to support:
> 
>     1. `~/.config/zed/settings.d/*.json` (e.g. applied as overrides, alphabetical order of filename?)

This is a well-established paradigm and is one I was just looking through the docs to see about. For my particular case, this would work well as my dotfiles are handled through Nix/home-manager and the files it creates are read only. I could create a `~/.config/zed/settings.d/nix-settings.json`

> 
>     2. Specifying an explicit settings.json at runtime -- e.g. `zed --config x/z/settings.json` (this would not support running two settings at the same time)
> 

This would probably also be a good feature, though it's less good for my particular case. It need not be a feature implemented instead of the above, but it might be good *in addition to* it. A second command line option could also be `--supplemental-config-dir x/z/settings.d/` that would allow the preceding folder from option 1 to be configurable.

> 
> A real issue for the practicality of any potential solution is the fact that settings.json currently includes a mix of "settings" (available_models, auto_update, ui_font_size, etc) which rarely change and various bits of state which are persisted via settings.json but can also be directly manipulated via Zed actions (theme, ssh_servers, default_model, vim_mode, base_keymap).
> 
> As soon as you have multiple files it's unclear where the latter stuff should go if you change it in the UI. In the `work_settings.json` example above, it would hard for Zed to know that editing certain servers should be in one file, but others would be in another (remember, you merged the configs based on order). Which also leads to the other issue, which is the complexity of merging multiple json files. Simple keys are easy (decide some order, and deterministically apply multiple files) but things get much harder with merging arrays. Sometimes you want to merge the arrays, sometimes order matters, sometimes it doesn't, do you need to dedupe values? sometimes!; and sometimes you shouldn't merge. Also applies at arbitrary JSON depths (lsp settings, etc).
> 
> Open to design ideas here, but nothing obvious jumps out at me to address all the desired use cases.

My proposed idea would be to implement both 1 and 2 above with the following defined behaviors:

1. Settings in the supplemental config directory would overwrite settings in the main config file.
2. If a non-default config file is passed as a CLI option, then no supplemental settings would be loaded by default.
3. If both a non-default config file AND a supplemental config directory are specified, then both are honored with the same behaviors as 1.
4. Keys would be merged from the bottom up, unless a parent key is explicitly set to `null` in a later file. e.g. if the primary config file holds `{ "foo": { "bar": {"baz": 1} } }` and a later file sets `foo.bar.also_baz` then both `baz` and `also_baz` are set in the final merged config. However, if the later file sets `{"foo": {"bar": null } }`, then all `foo.bar.*` from the earlier config files are effectively deleted.
5. When the settings are modified through the UI, only settings in the primary config file are modified. If these are subsequently overridden by files in the supplemental config directory, a message could be displayed to the user (perhaps as a tooltip or other warning) letting them know that currently that setting is affected by overriding configurations.

I don't have a fully formulated suggestion for merging arrays. That almost seems like it would need to have different behavior for each individual setting. Like, my SSH connections should probably be merged. But if I wanted to specify the order of commands to run for something, then it is less desirable.

Just a spitball idea: what about in each supplemental config file, you could specify something like `{ "meta": {"array_combine_strategy": "append|prepend|replace"} }` and it would take the specified action with array configurations discovered in that particular file? Default could be something seemingly appropriate, like appending lists, but then each file could override that behavior if it wanted?

> 
> See also:
> 
>     * [System wide config file for Zed: e.g. /etc/zed/ #22612](https://github.com/zed-industries/zed/issues/22612)

I acknowledge that this would be extremely non-trivial, but in the absence of anyone else replying with a suggested strategy, I figured I would throw out my ideas to hopefully spark the conversation!

**Comment by adreasnow** [comment]
I can't see it documented anywhere but in Zed there's the "switch profiles" command in the command palette, and also if you click on your account picture in the top-right of thew window (`ctrl`+`opt`+`cmd`+`P`).

This seems to correlate in `settings.zed` to a block like this:
```json
"profiles": {
  "Home": {},
  "Work": {}
}
```

And switching profiles seems to apply the settings in those blocks 🙂

**Comment by lentzi90** [comment]
This is still relevant.

**Comment by zelenenka** [comment]
@lentzi90 yep, sorry for the noise, the stalebot was overly eager due to this bug: https://github.com/actions/stale/issues/1302

**Comment by oldirtty** [comment]
'd love to see this implemented. Having a way to modularize/include settings would make managing dotfiles so much cleaner for shared and local settings.
